### PR TITLE
Tenant size calculation: refactor, rewrite, and add SVG

### PR DESCRIPTION
**Pull request by hlinnaka**
Refactor the tenant_size_model code. Segment now contains just the minimum amount of information needed to calculate the size. Other information that is useful for building up the segment tree, and for display purposes, is now kept elsewhere. The code in 'main.rs' has a new ScenarioBuilder struct for that.

Calculating which Segments are "needed" is now the responsibility of the caller of tenant_size_mode, not part of the calculation itself. So it's up to the caller to make all the decisions with retention periods for each branch.

The output of the sizing calculation is now a Vec of SizeResults, rather than a tree. It uses a tree representation internally, when doing the calculation, but it's not exposed to the caller anymore.

Refactor the way the recursive calculation is performed.

Rewrite the code in size.rs that builds the Segment model.  Get rid of the intermediate representation with Update structs. Build the Segments directly, with some local HashMaps and Vecs to track branch points to help with that.

retention_period is now an input to gather_inputs(), rather than an output.

Update pageserver http API: rename /size endpoint to /synthetic_size with following parameters:
    - /synthetic_size?inputs_only to get debug info;
    - /synthetic_size?retention_period=0 to override cutoff that is used to calculate the size;
    pass header  -H "Accept: text/html" to get HTML output, otherwise JSON is returned

Update python tests and openapi spec.

### CODE REVIEW

File: pageserver/src/tenant/size.rs [hunk 1]
```diff
@@ -37,13 +62,19 @@ struct TimelineInputs {
...
+
+    /// Cutoff point calculated from the user-supplied 'max_retention_period'
+    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
+    retention_param_cutoff: Option<Lsn>,
 }
 
 /// Gathers the inputs for the tenant sizing model.
 ///
-/// Tenant size does not consider the latest state, but only the state until next_gc_cutoff, which
+/// Tenant size does not consider the latest state (???), but only the state until next_gc_cutoff, which
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:77]: Before #2781 I don't see how it would had reflected the latest state, by which I meant logical size at `last_record_lsn`, which I had in the gathering originally and in discussions it was expected to be there (esp. in relation to early `last_record_lsn < gc_cutoff` tenants).

File: pageserver/src/tenant/size.rs [hunk 2]
```diff
@@ -247,104 +374,71 @@ pub(super) async fn gather_inputs(
...
+        }
+    }
+    Ok(())
 }
 
 impl ModelInputs {
-    pub fn calculate(&self) -> anyhow::Result<u64> {
+    pub fn calculate_model(&self) -> anyhow::Result<tenant_size_model::StorageModel> {
         // Option<TimelineId> is used for "naming" the branches because it is assumed to be
         // impossible to always determine the a one main branch.
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:412]:
```suggestion
```

Completly unrelated anymore.

File: pageserver/src/tenant/size.rs [hunk 3]
```diff
@@ -59,158 +90,260 @@ struct TimelineInputs {
...
+    let timeline_hash: HashMap<TimelineId, Arc<Timeline>> = HashMap::from_iter(
+        timelines
+            .iter()
+            .map(|timeline| (timeline.timeline_id, Arc::clone(timeline))),
+    );
+
+    // record the used/inserted cache keys here, to remove extras not to start leaking
+    // after initial run the cache should be quite stable, but live timelines will eventually
+    // require new lsns to be inspected.
+    let mut sizes_needed = HashMap::<(TimelineId, Lsn), Option<u64>>::new();
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/size.rs:312]: Nevermind, failed to see line 341. Resolving.

**Comment by hlinnaka** [comment]
Rebased and worked some more on this.

In some cases, we are getting different results with this PR, compared to what's in 'main'. This PR makes it easier to reason about the calculation, and check that it's correct, and when I looked at the tests that broke because the new calculation produced different results, I concluded that the new calculation is correct. However, I have not debugged why exactly the old code produced different results. At a high level, the logic should not have changed, so I believe the current code in 'main' has bugs that this fixes. It bothers me a little that I don't know what those bugs are, but not enough to actually spend the time debugging them.

1. One such case is 'test_single_branch_get_tenant_size_grows'. That's still failing, I didn't change the "expected output" of that. In a nutshell, it does this:

INSERT some rows
INSERT some rows
INSERT some rows
UPDATE some rows
UPDATE some rows
UPDATE some rows

And between each step, it checks that the synthetic size grows. However, I don't think that's correct, and with this PR, the synthetic size stops growing after the first UPDATE. The reason is that the test has a fixed GC horizon, and the UPDATEs merely move forward the amount of WAL that needs to be retained. Graphically:
After first UPDATE:
```
------============>
    ^   ^   ^   ^
    |   |   |   |
   INS INS INS UPD
```

After second UPDATE:
```
----------============>
    ^   ^   ^   ^   ^
    |   |   |   |   |
   INS INS INS UPD UPD
```
However, the INSERTs are not very large. Each INSERT only inserts 100 rows. If they all fit on a single page, then the INSERT doesn't grow the logical database size at all. In that case, the total synthetic size is exactly the same after the second UPDATE, as it was after the first UPDATE. In both cases, it consists of the logical size at beginning of the GC horizon, plus the WAL in the GC horizon. The GC horizon is a fixed amount of bytes in this tests. So if the logical size didn't change between the INSERTs, the synthetic size doesn't change either.

2. I think there was an off-by-one kind of bug in the the unit test scenarios. Scenario 1, for example, previously reported 8000, but believe 6000 is correct. Similarly, scenario 2 previously reported 9000, now 7000. I'm not entirely sure if this was a bug in the tests, or in the code.

3. Scenario 6 in the unit tests now produce a very different result. It was about 333 MB previously, now 136 MB. Looking at the new calculation and the dump of the model and the calculation, I believe the new result is corret. But I have no idea why the old code got a different result.

**Comment by hlinnaka** [comment]
There are some FIXMEs and a test failure, but I'd like to get some review of this. Do the new StorageModel and SizeResult representations make sense? What about the code that builds the StorageModel in pageserver/src/tenant/size.rs?

**Comment by hlinnaka** [comment]
> 1. One such case is 'test_single_branch_get_tenant_size_grows'. That's still failing, I didn't change the "expected output" of that. In a nutshell, it does this:

I hacked the test to print out the model and the size calculation result after each INSERT and UPDATE batch. See [`test_output/test_single_branch_get_tenant_size_grows/size_debug.html`](https://gist.github.com/hlinnaka/98bc5e1ff8d4e49ffbe97ba4b4e9c1af) after running the test.

File: test_runner/regress/test_tenant_size.py [hunk 4]
```diff
@@ -7,6 +9,34 @@
 from fixtures.types import Lsn
 
 
+def mask_model_inputs(x):
```
Comment (koivunej) [inline test_runner/regress/test_tenant_size.py:12]: Could add a comment why this is needed (#2592?) and perhaps an example, or perhaps the example could be a test case, it's a bit difficult to see what does this do.

File: pageserver/src/tenant/size.rs [hunk 5]
```diff
@@ -604,227 +487,230 @@ async fn calculate_logical_size(
...
+      "next_gc_cutoff": "0/18B3D98",
+      "retention_param_cutoff": null
+    }
+  ]
+}
+"#;
     let inputs: ModelInputs = serde_json::from_str(doc).unwrap();
 
-    assert_eq!(inputs.calculate().unwrap(), 36_409_872);
+    assert_eq!(inputs.calculate().unwrap(), 37_851_408);
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:647]: Do you know why this changed?

File: pageserver/src/tenant/size.rs [hunk 6]
```diff
@@ -181,257 +120,265 @@ fn sort_updates_in_tree_order(updates: Vec<Update>) -> anyhow::Result<Vec<Update
...
+        let mut lsns: Vec<(Lsn, LsnKind)> = gc_info
+            .retain_lsns
+            .iter()
+            .inspect(|&&lsn| {
+                trace!(
+                    timeline_id=%timeline.timeline_id,
+                    "retained lsn: {lsn:?}, is_before_ancestor_lsn={}",
+                    lsn < ancestor_lsn
+                )
+            })
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:213]:
```suggestion
```

I don't see any help in keeping this anymore, often hoped to remove it but haven't had time to do any refactorings near this.

File: pageserver/src/tenant/size.rs [hunk 7]
```diff
@@ -181,257 +120,265 @@ fn sort_updates_in_tree_order(updates: Vec<Update>) -> anyhow::Result<Vec<Update
...
-                }
-                _ => {
-                    max_cutoff_distance = Some(cutoff_distance);
-                }
-            }
+        // the minimum where we should find the next_gc_cutoff for our calculations.
+        //
+        // next_gc_cutoff in parent branch are not of interest (right now at least), nor do we
+        // want to query any logical size before initdb_lsn.
+        let branch_start_lsn = cmp::max(ancestor_lsn, timeline.initdb_lsn);
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:201]: At first glance the comment seems unrelated now that the name is no longer `cutoff_minimum`, took me a while. `branch_start_lsn` though makes more sense. Cannot suggest a better comment however.

File: pageserver/src/tenant/size.rs [hunk 8]
```diff
@@ -181,257 +120,265 @@ fn sort_updates_in_tree_order(updates: Vec<Update>) -> anyhow::Result<Vec<Update
...
+            })
+            .filter(|&&lsn| lsn > ancestor_lsn)
+            .copied()
+            // this assumes there are no other retain_lsns than the branchpoints
+            .map(|lsn| (lsn, LsnKind::BranchPoint))
+            .collect::<Vec<_>>();
+
+        // Add branch points we collected earlier, just in case there were any that were
+        // not present in retain_lsns. Shouldn't happen, but better safe than sorry.
+        // We will remove any duplicates below later.
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:222]: This will naturally happen if the "main" branch is branched right after initdb_lsn and then continued in the branch beyond gc_horizon -- the main branch will never have it's gc_info refreshed, created `test_branched_empty_timeline_size` for this.

```suggestion
        // Add branch points we collected earlier, just in case there were any that were
        // not present in retain_lsns. We will remove any duplicates below later.
```

File: pageserver/src/tenant/size.rs [hunk 9]
```diff
@@ -1,44 +1,87 @@
...
+}
+
+#[derive(
+    Debug, Clone, Copy, Eq, Ord, PartialEq, PartialOrd, serde::Serialize, serde::Deserialize,
+)]
+pub enum LsnKind {
+    BranchStart, // this timeline starts here
+    BranchPoint, // a child timeline branches off from here
+    GcCutOff,    // GC cutoff point
+    BranchEnd,   // last record LSN
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:74]:
```suggestion
    /// A timeline starting here
    BranchStart,
    /// A child timeline branches off from here 
    BranchPoint,
    /// GC cutoff point
    GcCutOff,
    /// Last record LSN
    BranchEnd,
```

minor rustdocification.

File: pageserver/src/tenant/size.rs [hunk 10]
```diff
@@ -1,44 +1,87 @@
...
+    fn size_needed(&self) -> bool {
+        match self.kind {
+            LsnKind::BranchStart => {
+                // If we don't have a later GcCutoff point on this branch, and
+                // no ancestor, calculate size for the branch start point.
+                if self.segment.needed && self.segment.parent.is_none() {
+                    true
+                } else {
+                    false
+                }
```
Comment (koivunej) [inline pageserver/src/tenant/size.rs:58]:
```suggestion
                self.segment.needed && self.segment.parent.is_none()
```

File: pageserver/src/http/routes.rs [hunk 11]
```diff
@@ -519,14 +519,16 @@ async fn tenant_size_handler(request: Request<Body>) -> Result<Response<Body>, A
         ///
         /// Will be none if `?inputs_only=true` was given.
         size: Option<u64>,
+        segment_sizes: Option<Vec<tenant_size_model::SegmentSizeResult>>,
```
Comment (koivunej) [inline pageserver/src/http/routes.rs:522]: Should this be noted above in the type comment next to '"unstable" `inputs` field' or put into swagger desc?

File: pageserver/src/tenant.rs [hunk 12]
```diff
@@ -2418,6 +2418,7 @@ impl Tenant {
     #[instrument(skip_all, fields(tenant_id=%self.tenant_id))]
     pub async fn gather_size_inputs(
         &self,
+        max_retention_period: Option<u64>,
```
Comment (koivunej) [inline pageserver/src/tenant.rs:2421]: Add comment on how this is meant to be used? I think this will be used to try the pitr=0 difference, but unsure.

**Review by koivunej — COMMENTED** [review]
Is this waiting for some bigger changes (drafted)? Also, did you try through the perhaps duplicate skipped tests in test_tenant_size.py?

**Comment by lubennikovaav** [comment]
Quick question before I start review.
Does this PR follow the documentation in PR #3328?

File: libs/tenant_size_model/src/lib.rs [hunk 13]
```diff
@@ -1,401 +1,69 @@
...
-
-    /// Description of how did we get to this state.
-    ///
-    /// Mainly used in the original scenarios 1..=4 with insert, delete and update. Not used when
-    /// modifying a branch directly.
-    pub op: Cow<'static, str>,
-
-    /// LSN before this state
-    start_lsn: u64,
+    pub parent: Option<usize>,
```
Comment (duskpoet) [inline libs/tenant_size_model/src/lib.rs:34]: Hey! Very cool job on redefining synthetic size!
So I was figuring out possible ways to visualize branches sizes for users with @kelvich and @lubennikovaav , and we went through a lot of code and possible approaches and figured out that visualizing the segments.json as array of segments is the most straightforward approach now. For that reason, can we add `ancestor_timeline_id` to the segment?

**Comment by lubennikovaav** [comment]
I addressed all review issues.
The only remaining issue is `test_single_branch_get_tenant_size_grows`

File: pageserver/src/tenant/size.rs (hunk 8 above)
Comment (koivunej) [reply pageserver/src/tenant/size.rs]: Unsure if this was changed...?

Comment (koivunej) [reply pageserver/src/tenant/size.rs]: Yes it was, resolving again.

File: pageserver/src/http/routes.rs (hunk 11 above)
Comment (koivunej) [reply pageserver/src/http/routes.rs:525]: Unstable it is, no swagger desc yet, resolving.

File: pageserver/src/tenant/size.rs (hunk 7 above)
Comment (koivunej) [reply pageserver/src/tenant/size.rs]: Unsure how or if this was changed, but it was my minor note anyways. resolving.

File: pageserver/src/tenant/size.rs (hunk 6 above)
Comment (koivunej) [reply pageserver/src/tenant/size.rs]: This has been removed.

File: libs/tenant_size_model/src/lib.rs (hunk 13 above)
Comment (koivunej) [reply libs/tenant_size_model/src/lib.rs:34]: Was this fixed in [39bc04a](https://github.com/neondatabase/neon/pull/2817/commits/39bc04a5292558499fa6f3ae454405525fb89740)?

File: pageserver/src/tenant/size.rs (hunk 5 above)
Comment (lubennikovaav) [reply pageserver/src/tenant/size.rs:642]: I don't know, but it seems legit

File: test_runner/regress/test_tenant_size.py [hunk 14]
```diff
@@ -436,9 +450,16 @@ def test_single_branch_get_tenant_size_grows(
...
-            ), "even though rows have been deleted, the tenant_size should increase"
+
+            if current_lsn - initdb_lsn > gc_horizon:
+                assert (
+                    size >= prev
+                ), "tenant_size may grow or not grow, because we only add gc_horizon amount of WAL to initial snapshot size"
+            else:
+                assert (
+                    size > prev
+                ), "tenant_size should grow, because we continue to add WAL to initial snapshot size"
```
Comment (koivunej) [inline test_runner/regress/test_tenant_size.py:431]: Assuming this is a correct property which must hold, maybe refactor it to an test-function internal function which allows these to be replaced as:

```python
check_growth(current_lsn, prev, size)
```

`check_growth` is defined somewhere after `initdb_lsn` and `gc_horizon` have been defined:

```
def check_growth(current_lsn, prev_size, size_now):
  if current_lsn - initdb_lsn > gc_horizon:
    assert size_now >= prev_size, "long rationale"
  else:
    assert size_now > prev_size, "long rationale"
```

File: test_runner/regress/test_tenant_size.py [hunk 15]
```diff
@@ -467,16 +497,20 @@ def test_single_branch_get_tenant_size_grows(
...
 
-    tenant_size_metric = int(
-        ps_metrics.query_one("pageserver_tenant_synthetic_size", filter=tenant_metric_filter).value
-    )
+    # tenant_size_metric = int(
+    #     ps_metrics.query_one("pageserver_tenant_synthetic_size", filter=tenant_metric_filter).value
+    # )
 
-    assert tenant_size_metric == size_after, "API size value should be equal to metric size value"
+    # assert tenant_size_metric == size_after, "API size value should be equal to metric size value"
```
Comment (koivunej) [inline test_runner/regress/test_tenant_size.py:483]: Could this be moved to consumption metrics testing?

Comment (lubennikovaav) [reply test_runner/regress/test_tenant_size.py:483]: I am inclined to remove this at all. This code tests prometheus metric, which now simply mirrors the consumption metric.

If anything goes wrong, it will be easier to notice it in Grafana, than in python test.

Comment (koivunej) [reply test_runner/regress/test_tenant_size.py:483]: I am fine with that.

File: pageserver/src/tenant/size.rs (hunk 5 above)
Comment (koivunej) [reply pageserver/src/tenant/size.rs:642]: Yes, sounds good enough. Resolving.

**Comment by lubennikovaav** [comment]
New API proposal:

`/size` calculates size and returns it a short json

`/size?inputs_only` doesn't calculate size, returns raw inputs

`/size?retention_period` calculates size, as if gc_horizon = retention_period
returns size + verbose json with segments & timeline_inputs

`/size?svg` returns html with SVG and verbose json
This will be removed as soon as we proper visualization.

`svg` and `inputs_only` are mutually exclusive

all parameters are debug only and won't be mentioned in openAPI spec.

@hlinnaka , @koivunej , WDYT?

**Comment by kelvich** [comment]
Also since we settled on `synthetic_size` terminology, may be also change `/size` to `/synthetic_size`?

**Comment by koivunej** [comment]
> I suggest that we unify HTTP API endpoints.

retention period would probably be useful in the `/size` or `/synthetic_size` as well (needs an openapi adjustment).

Requester's `Accept` header could be inspected if they want an svg or json, would that work? Default on json, and return 406 on something else?

**Comment by hlinnaka** [comment]
> New API proposal:

+1

> `/size` calculates size and returns it a short json
> 
> `/size?inputs_only` doesn't calculate size, returns raw inputs
> 
> `/size?retention_period` calculates size, as if gc_horizon = retention_period returns size + verbose json with segments & timeline_inputs
> 
> `/size?svg` returns html with SVG and verbose json This will be removed as soon as we proper visualization.
>
> `svg` and `inputs_only` are mutually exclusive

Hmm, I guess you could print the SVG without calculating the sizes too. It would be just tree of inputs then. We don't really need it, though, so no need to spend much effort on it if it doesn't work naturally.

> all parameters are debug only and won't be mentioned in openAPI spec.

There has been discussion on showing the size without history, i.e. with `retention_period=0`, in the console too.

+1 to @kelvich 's suggest for calling it `synthehtic_size`. All of the options, svg, retention_period, inputs_only, seem orthogonal. I suggest that we make them all independently settable:

**Comment by duskpoet** [comment]
> I suggest that we unify HTTP API endpoints.
> @duskpoet , what output format will be more convenient for you? plain json, I suppose?

so does /size returns an array of segments now, as json? That should suffice for the start

**Review by koivunej — APPROVED** [review]
